### ISSUE TITLE
Uncaught SyntaxError: Unexpected token u in JSON at position 0

### ISSUE THREAD

**Issue by m-nathani** [kilkelly/redux-localstorage-simple#24]
json parse error while providing the following preloadedstate.. 

```
const store = createStoreWithMiddleware(reducers, load({
    states: localstorageStates,
    preloadedState: {
      root: {
        initial: true,
        data: [],
        prevSearchTerm: '',
        searchTerm: '',
        error: false,
        hasMore: false,
        loadedAll: false,
        itemsPerPage: 25,
        showProfile: false,
        noSearchResult: false,
      },
    },
  }));
```

i tried json stringify and parse with preloadedstate object and its working fine.

Thanks

**Comment by kilkelly** [comment]
Hi @m-nathani , this error can happen when a localstorage key exists for a given state but it contains an `undefined` value. It can happen sometimes during development. I notice from your [code](https://github.com/m-nathani/giphy-love/blob/04cc2831e197681acb5b890829c78c1d23a71bb4/src/store/index.js#L28) that you are not using JSON.stringify/parse. Did `preloadedState` work as intended eventually?

**Comment by m-nathani** [comment]
Hey kilkelly,   yes preloadedstate worked. But the json breaks even if i remove any attribute from the root object that is present in my initialState. 

So If i do not provide data that's not in preloadedstate it should set the data from the reducers intialState.

**Comment by m-nathani** [comment]
Hey kilkelly,   yes preloadedstate worked. But the json breaks even if i remove any attribute from the root object that is present in my initialState. 

So If i do not provide data that's not in preloadedstate it should set the data from the reducers intialState.

**Comment by kilkelly** [comment]
Do you have the initialState contents that I could take a look at?

**Comment by m-nathani** [comment]
> Do you have the initialState contents that I could take a look at?

Below is the link for the initialState in reducer. Currently i have matched it exactly as preloadedstate. any change in preloadedstate is breaking it. 
[reducer](https://github.com/m-nathani/giphy-love/blob/master/src/reducer/root.js)

**Comment by kilkelly** [comment]
Thanks. I notice that you specify states like `root.favorites` and `root.isLoading` for redux-localstorage-simple, however in `initialState` the `root` property is missing. Can you test it with the parent `root` property added?

**Comment by m-nathani** [comment]
> Thanks. I notice that you specify states like `root.favorites` and `root.isLoading` for redux-localstorage-simple, however in `initialState` the `root` property is missing. Can you test it with the parent `root` property added?

@kilkelly , the following code is the reason why i have to wrap my `initialState` with `root` . as while using multiple reducers i need to combine them using `combineReducers` . 

[reducers](https://github.com/m-nathani/giphy-love/blob/master/src/reducer/index.js)

```
export default combineReducers({
  routing,
  root,
});
```

**Comment by kilkelly** [comment]
Hi @m-nathani , I've just fixed a bug to prevent the JSON error. You will now receive a warning in the console if an invalid state has been specified. This is the original intended behaviour, however my check for `undefined` in localStorage was bugged. Upgrade to 2.1.5 and let me know what you observe now.

**Comment by glebtv** [comment]
Still does not work with the same error with ```load({states: ["some states"], namespace: "some]})``` and empty localstorage
```load({namespace: "some"})``` without ```states: []``` works

This check seems to still be bugged, probably should be ```typeof localStorage[...] !== "undefined"```
https://github.com/kilkelly/redux-localstorage-simple/blob/master/dist/index.js#L326

**Comment by kilkelly** [comment]
Alright all, seems I finally got around to the root of this error. When an incorrect state was specified it was being created in LocalStorage with the value `undefined` when it should be created at all. For example the states:
```

  let states = [
    'validstate',
    'invalidstate' // this state doesn't exist in the app so shouldn't be added to LocalStorage
  ]


```
would create `invalidstate` in LocalStorage with the value `undefined`. This is what my code was checking for. However I've now made a change so this invalid state is never added to LocalStorage. 

Would love some feedback if you have tried it and it worked for you 👍
